### DISCUSSION TITLE
return the default value when json is empty

### DISCUSSION THREAD

**Discussion by BKSpek** [nlohmann/json#4415] [Ideas]
Using the value function on an empty json throws an exception. I propose simply returning the default value because ultimately (for me at least) that seems like it would be the appropriate behavior for not finding the value requested.

For anyone wondering how to currently work around the thrown exception, assign a nlohmann::json::initializer_list_t to the json object before calling value.

```
ValueType value(const typename object_t::key_type& key, const ValueType& default_value) const
{
    if (is_null()) 
    {
        return default_value;
    }

... rest of the function
```

**Comment by nlohmann** [comment]
The documentation (https://json.nlohmann.me/api/basic_json/value) states:

> Throws [type_error.306](https://json.nlohmann.me/home/exceptions/#jsonexceptiontype_error306) if the JSON value is not an object; in that case, using value() with a key makes no sense.

Your change would break this contract, so we unfortunately cannot make it.

**Comment by gregmarr** [comment]
If the requested value is `null`, then it's an error.  That's different than the value not being found, which would indeed return the default value.  Do you have code that shows different behavior?

**Comment by BKSpek** [comment]
In my mind the default value would be returned in the event of a failure. But I can see the importance in the distinction between not found and invalid request, especially in the context of lists and literal types. Thank you for the prompt feedback.
